src/jsonrpc/server.rs

```diff
@@ -0,0 +1,429 @@
+            // request" or "invalid JSON" error.
+            let req: Request = serde_json::from_slice(&buf[..*cursor + pos])?;
+            *buf = buf[pos + 1..].to_vec(); // FIXME: can we avoid reallocating here?
+            *cursor = 0;
+            *end -= pos + 1;
+
+            return Ok(Some(req));
+        }
+
+        // If nothing can be gathered from the buffer, continue reading.
+        let new_read = stream.read(&mut buf[*end..])?;
+        if new_read == 0 {
+            return Ok(None);
+        }
+
+        // If we filled the buffer, increase its size and try again.
+        *end += new_read;
+        let buffer_filled = *end == buf.len();
+        if buffer_filled {
+            buf.resize(buf.len() * 2, 0);
+            continue;
+        }
+    }
+}
+
+// Handle all messages from this connection.
+fn connection_handler(
+    control: DaemonControl,
+    mut stream: net::UnixStream,
+    shutdown: sync::Arc<atomic::AtomicBool>,
+) -> Result<(), io::Error> {
+    let mut buf = vec![0; 2048];
+    let mut end = 0;
+    let mut cursor = 0;
+
+    while !shutdown.load(atomic::Ordering::Relaxed) {
+        let req = match read_command(&mut stream, &mut buf, &mut end, &mut cursor)? {
+            Some(req) => req,
+            None => {
+                // Connection closed.
+                return Ok(());
+            }
+        };
+
+        let req_id = req.id.clone();
+        if &req.method == "stop" {
+            shutdown.store(true, atomic::Ordering::Relaxed);
+            log::info!("Stopping the minisafe daemon.");
+        }
+
+        log::trace!("JSONRPC request: {:?}", serde_json::to_string(&req));
+        let response =
+            api::handle_request(&control, req).unwrap_or_else(|e| Response::error(req_id, e));
+        log::trace!("JSONRPC response: {:?}", serde_json::to_string(&response));
+        if let Err(e) = serde_json::to_writer(&stream, &response) {
+            log::error!("Error writing response: '{}'", e);
+            return Ok(());
+        }
+    }
+
+    Ok(())
+}
+
+// FIXME: have a decent way to share the DaemonControl between connections. Maybe make it Clone?
+/// The main event loop. Wait for connections, and treat requests sent through them.
+pub fn rpcserver_loop(
+    listener: net::UnixListener,
+    daemon_control: DaemonControl,
+) -> Result<(), io::Error> {
+    // Keep it simple. We don't need great performances so just treat each connection in
+    // its thread, with a given maximum number of connections.
+    let connections_counter = sync::Arc::from(atomic::AtomicU32::new(0));
+    let shutdown = sync::Arc::from(atomic::AtomicBool::new(false));
+
+    listener.set_nonblocking(true)?;
+    while !shutdown.load(atomic::Ordering::Relaxed) {
+        let (connection, _) = match listener.accept() {
+            Ok(c) => c,
+            Err(e) => {
+                log::error!("Accepting new connection: '{}'", e);
+                continue;
```

Comment:
Since the socket is set to nonblocking: don't log as error here, and sleep a few milliseconds.